pkg/epp/scheduling/scheduler.go

```diff
@@ -67,89 +65,83 @@ func LoadConfig() Config {
+					simpleFilter: leastKVCacheFilter,
+				},
+			},
 		},
-		nextOnFailure: queueLoRAAndKVCacheFilter,
 	}
 
 	sheddableRequestFilter = &filter{
 		// When there is at least one model server that's not queuing requests, and still has KV
 		// cache below a certain threshold, we consider this model server has capacity to handle
 		// a sheddable request without impacting critical requests.
-		name:          "has capacity for sheddable requests",
-		filter:        toFilterFunc(noQueueAndLessThanKVCacheThresholdPredicate(config.QueueThresholdCritical, config.KVCacheThreshold)),
-		nextOnSuccess: queueLoRAAndKVCacheFilter,
+		simpleFilter: simpleFilter{
+			name:   "has capacity for sheddable requests",
+			filter: toFilterFunc(queueThresholdPredicate(config.QueueThresholdCritical).and(kvCacheThresholdPredicate(config.KVCacheThreshold))),
+		},
+		nextOnSuccess: lowLatencyFilter,
 		// If all pods are queuing or running above the KVCache threshold, we drop the sheddable
 		// request to make room for critical requests.
-		nextOnFailure: &filter{
+		nextOnFailure: &simpleFilter{
 			name: "drop request",
-			filter: func(logger logr.Logger, req *LLMRequest, pods []backendmetrics.PodMetrics) ([]backendmetrics.PodMetrics, error) {
-				logger.V(logutil.DEFAULT).Info("Request dropped", "request", req)
-				return []backendmetrics.PodMetrics{}, errutil.Error{
+			filter: func(ctx *Context, pods []*PodMetrics) ([]*PodMetrics, error) {
+				ctx.logger.V(logutil.DEFAULT).Info("Request dropped", "request", ctx.req)
+				return []*PodMetrics{}, errutil.Error{
 					Code: errutil.InferencePoolResourceExhausted, Msg: "dropping request due to limited backend resources",
 				}
 			},
 		},
 	}
 )
 
-func NewScheduler(datastore datastore.Datastore) *Scheduler {
+func NewScheduler(datastore Datastore) *Scheduler {
 	return &Scheduler{
 		datastore: datastore,
-		filter:    defaultFilter,
 	}
 }
 
 type Scheduler struct {
-	datastore datastore.Datastore
-	filter    Filter
+	datastore Datastore
+}
+
+type Datastore interface {
+	PodGetAll() []backendmetrics.PodMetrics
 }
 
 // Schedule finds the target pod based on metrics and the requested lora adapter.
-func (s *Scheduler) Schedule(ctx context.Context, req *LLMRequest) (targetPod backendmetrics.PodMetrics, err error) {
+func (s *Scheduler) Schedule(ctx context.Context, req *LLMRequest) (targetPod Pod, err error) {
 	logger := log.FromContext(ctx).WithValues("request", req)
 
-	podMetrics := s.datastore.PodGetAll()
-	logger.V(logutil.DEBUG).Info(fmt.Sprintf("Scheduling a request. Metrics: %+v", podMetrics))
+	// Snapshot pod metrics from the datastore to:
+	// 1. Reduce concurrent access to the datastore.
+	// 2. Ensure consistent data during the scheduling operation of a request.
+	sCtx := newContext(ctx, req, toSchedulerPodMetrics(s.datastore.PodGetAll()))
+	logger.V(logutil.DEBUG).Info(fmt.Sprintf("Scheduling a request. Metrics: %+v", sCtx.podsSnapshot))
+
+	var filter Filter
+	if req.Critical {
```

Comment:
It's much cleaner this way to show how we handle critical vs. sheddable differently, than the previous filter.